### PR TITLE
Encrypts getBalance and call requests to the enclave from the wallet extension

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

https://github.com/obscuronet/obscuro-internal/issues/200

### What changes were made as part of this PR:

Functional.

- Encrypts getBalance requests to enclave
- Encrypts call requests to enclave

### CODE REVIEW

File: tools/walletextension/wallet_extension.go [hunk 1]
```diff
@@ -119,17 +120,31 @@ func (we *WalletExtension) handleHTTPEthJSON(resp http.ResponseWriter, req *http
...
-	//if err != nil {
-	//	logAndSendErr(resp, fmt.Sprintf("could not encrypt request with enclave public key: %s", err))
-	//	return
-	//}
+	// We encrypt the request's params with the enclave's public key if it's a sensitive request.
+	maybeEncryptedBody := body
+	if isSensitive(method) {
+		fmt.Println("🔒 Encrypting request from wallet with enclave public key.")
+		params := reqJSONMap[reqJSONKeyParams]
+		paramsJSON, err := json.Marshal(params)
```
Comment (BedrockSquirrel) [inline tools/walletextension/wallet_extension.go:128]: Again there's a lot of low-level fiddling here and the method becomes quite hard to follow (getting towards 100 lines long).

Maybe we could break it up into some high-level chunks that hide away the details for readability, like something like this could be easier to scan for the reader maybe:
```
fwdReq, err = encryptRequestIfNeeded(req.body)

resp, err = forwardRequestAndAwaitResponse(fwdReq)

clientResp, err = decryptResponseIfNeeded(resp)

returnResponse(clientResp)
```

File: go/obscuronode/nodecommon/types.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ type (
 	L2Tx                  = types.Transaction
 	EncryptedTx           []byte // A single transaction encrypted using the enclave's public key
 	EncryptedTransactions []byte // A blob of encrypted transactions, as they're stored in the rollup.
+	EncryptedParams       []byte // The encrypted params for an off-chain call or getBalance request. Encrypted with the public key of the enclave.
```
Comment (BedrockSquirrel) [inline go/obscuronode/nodecommon/types.go:22]: Perhaps worth expanding here and/or on the enclave interface what this blob contains so reader doesn't have to go find the usages to imagine it (like saying it's json key-val pairs and potentially even listing what's in there since there's not many)

**Review by BedrockSquirrel — COMMENTED** [review]
Few minor (and probably subjective) comments

File: go/obscuronode/host/in_mem_obscuro_client.go [hunk 3]
```diff
@@ -84,23 +82,15 @@ func (c *inMemObscuroClient) Call(result interface{}, method string, args ...int
 		*result.(**nodecommon.L2Tx) = c.obscuroAPI.GetTransaction(hash)
 
 	case obscuroclient.RPCCall:
-		if len(args) != 3 {
+		if len(args) != 1 {
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/obscuronode/host/in_mem_obscuro_client.go:85]: :+1

File: go/obscuronode/nodecommon/types.go (hunk 2 above)
Comment (joeldudley) [reply go/obscuronode/nodecommon/types.go:22]: :+1

File: go/obscuronode/enclave/enclave.go [hunk 4]
```diff
@@ -402,7 +410,44 @@ func (e *enclaveImpl) notifySpeculative(winnerRollup *obscurocore.Rollup) {
...
+	paramBytes := encryptedParams
+	if e.config.ViewingKeysEnabled {
+		var err error
+		paramBytes, err = ecies.ImportECDSA(e.privateKey).Decrypt(encryptedParams, nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("could not decrypt params in Call request. Cause: %w", err)
+		}
+	}
+
+	var paramsJSONMap []interface{}
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/obscuronode/enclave/enclave.go:424]: I agree but I didn't want to start doing refactoring here. Will do in a follow-up?

File: tools/walletextension/wallet_extension.go (hunk 1 above)
Comment (joeldudley) [reply tools/walletextension/wallet_extension.go:128]: I agree but I didn't want to start doing refactoring here. Will do in a follow-up?

**Review by BedrockSquirrel — APPROVED** [review]
LGTM
